### PR TITLE
Big int

### PR DESCRIPTION
**Pull request by ArnabChatterjee20k**
## Summary by CodeRabbit

* **New Features**
  * Added support for a BIGINT attribute type for 64-bit signed/unsigned integer storage and consistent handling across adapters and queries.
* **Bug Fixes**
  * Improved type mapping and validation so bigint values are stored, cast, and queried reliably and size accounting is adjusted where applicable.
* **Tests**
  * Added unit and end-to-end tests covering BIGINT attribute creation, validation, document operations, increments/decrements, and boundary values.

### CODE REVIEW

File: src/Database/Adapter/SQL.php [hunk 1]
```diff
@@ -896,6 +896,20 @@ public function getLimitForInt(): int
...
+     * Get max BIGINT limit
+     *
+     * @return int
+     */
+    public function getLimitForBigInt(): int
+    {
+        // 2^64 - 1
+        // 18446744073709551615 is the maximum value for a 64-bit unsigned integer
+        // 9223372036854775807 is the maximum value for a 64-bit signed integer
+        // in php we can't represent 64-bit integer, so greater than 4294967295 will be treated as bigint
```
Comment (ArnabChatterjee20k) [inline src/Database/Adapter/SQL.php]: @abnegate in php no option for representing a bigint in form of int so having 18446744073709551615 we need to consider string for precision but we can't store the size incoming from user inside the metadata as it will be string and we have to start taking input as string
Shall we go forward with string ? or BIGINT will be marking in the table only?

File: src/Database/Database.php [hunk 2]
```diff
@@ -42,6 +42,7 @@ class Database
     // Simple Types
     public const VAR_STRING = 'string';
     public const VAR_INTEGER = 'integer';
+    public const VAR_BIGINT = 'bigint';
```
[the remark this answers is not in the document]
Comment (abnegate) [reply src/Database/Database.php]: @ArnabChatterjee20k Comment is valid

File: src/Database/Adapter/MariaDB.php [hunk 3]
```diff
@@ -2251,6 +2255,11 @@ public function getSupportForObject(): bool
         return false;
     }
 
+    public function getSupportForUnsignedBigInt(): bool
```
Comment (ArnabChatterjee20k) [inline src/Database/Adapter/MariaDB.php:2280]: mongodb, postgresql, sqlite doesn't support unsigned bigint

File: src/Database/Adapter/SQL.php [hunk 4]
```diff
@@ -896,6 +896,16 @@ public function getLimitForInt(): int
...
     }
 
+    /**
+     * Get max BIGINT limit
+     *
+     * @return int
+     */
+    public function getLimitForBigInt(): int
+    {
+        return Database::MAX_BIG_INT;
```
Comment (ArnabChatterjee20k) [inline src/Database/Adapter/SQL.php:922]: Here going other way around
It is the signed big int limit(supported by all adapter)
For adapters supporting unsigned int, in the attribute validator, we are validating via string comparison

File: src/Database/Adapter/Mongo.php [hunk 5]
```diff
@@ -1319,6 +1319,7 @@ public function castingAfter(Document $collection, Document $document): Document
             foreach ($value as &$node) {
                 switch ($type) {
                     case Database::VAR_INTEGER:
+                    case Database::VAR_BIGINT:
                         $node = (int)$node;
                         break;
```
[the remark this answers is not in the document]
Comment (ArnabChatterjee20k) [reply src/Database/Adapter/Mongo.php:1335]: Mongodb doesn't goes out of the php safe int limit and doesn't support unsigned int. So same as integer only

File: src/Database/Helpers/BigInt.php [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        return \preg_match($signed ? '/^-?\d+$/' : '/^\d+$/', $value) === 1;
+    }
+
+    public static function fitsPhpInt(string $value, bool $signed = true): bool
+    {
+        if (!self::isIntegerString($value, $signed)) {
+            return false;
+        }
+
+        $phpMax = (string)\PHP_INT_MAX;
```
Comment (abnegate) [inline src/Database/Helpers/BigInt.php]: Let's move all of this class into the validator

File: src/Database/Validator/Query/Filter.php [hunk 7]
```diff
@@ -155,12 +157,17 @@ protected function isValidAttributeAndValues(string $attribute, array $values, s
                 case Database::VAR_INTEGER:
                     $size = $attributeSchema['size'] ?? 4;
                     $signed = $attributeSchema['signed'] ?? true;
-                    $bits = $size >= 8 ? 64 : 32;
+                    $bits = ($attributeType === Database::VAR_BIGINT || $size >= 8) ? 64 : 32;
```
Comment (abnegate) [inline src/Database/Validator/Query/Filter.php]: The check is redundant, type is always integer here

File: src/Database/Validator/Attribute.php [hunk 8]
```diff
@@ -522,6 +551,12 @@ protected function validateDefaultTypes(string $type, mixed $default): void
                     throw new DatabaseException($this->message);
                 }
                 break;
+            case Database::VAR_BIGINT:
+                if ($defaultType !== 'integer') {
+                    $this->message = 'Default value ' . $default . ' does not match given type ' . $type;
+                    throw new DatabaseException($this->message);
+                }
```
[the remark this answers is not in the document]
Comment (abnegate) [reply src/Database/Validator/Attribute.php:543]: @ArnabChatterjee20k Seems valid?

File: src/Database/Validator/Structure.php [hunk 9]
```diff
@@ -353,16 +364,21 @@ protected function checkForInvalidAttributeValues(array $structure, array $keys)
 
                 case Database::VAR_INTEGER:
                     // Determine bit size based on attribute size in bytes
-                    $bits = $size >= 8 ? 64 : 32;
+                    // BIGINT is always 64-bit in SQL adapters; VAR_INTEGER uses size to decide.
+                    $bits = ($type === Database::VAR_BIGINT || $size >= 8) ? 64 : 32;
                     // For 64-bit unsigned, use signed since PHP doesn't support true 64-bit unsigned
```
Comment (abnegate) [inline src/Database/Validator/Structure.php:368]: Same here, type can't be bigint if we're in the int case

File: src/Database/Database.php [hunk 10]
```diff
@@ -2588,6 +2592,11 @@ protected function validateDefaultTypes(string $type, mixed $default): void
                     throw new DatabaseException('Default value ' . $default . ' does not match given type ' . $type);
                 }
                 break;
+            case Database::VAR_BIGINT:
+                if ($defaultType !== 'integer') {
```
Comment (abnegate) [inline src/Database/Database.php]: Do we need string here too?

File: src/Database/Database.php [hunk 11]
```diff
@@ -5577,7 +5598,9 @@ public function createDocument(string $collection, Document $document): Document
                 $this->adapter->getIdAttributeType(),
                 $this->adapter->getMinDateTime(),
                 $this->adapter->getMaxDateTime(),
-                $this->adapter->getSupportForAttributes()
+                $this->adapter->getSupportForAttributes(),
+                null,
```
Comment (abnegate) [inline src/Database/Database.php]: Do we need to pass the adapter values here

File: src/Database/Validator/Attribute.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (ArnabChatterjee20k) [reply src/Database/Validator/Attribute.php:543]: Its already present(https://github.com/utopia-php/database/pull/847/changes#diff-6b7714b772da21b04d82d9ec7e003ce1155773a838e8ddf53f96f51415c5235bR538-R540), not sure why its still showing

File: src/Database/Validator/Structure.php (hunk 9 above)
Comment (ArnabChatterjee20k) [reply src/Database/Validator/Structure.php:368]: yep, was a leftover from the revert , updating

File: src/Database/Database.php (hunk 11 above)
Comment (ArnabChatterjee20k) [reply src/Database/Database.php]: yes cause mongodb, postgres, sqlite doesn't support unsigned bigint
So for the structure validator we needed this
passing null here as current document by default is null

File: src/Database/Validator/Structure.php (hunk 9 above)
Comment (ArnabChatterjee20k) [reply src/Database/Validator/Structure.php:368]: updated

File: src/Database/Validator/Query/Filter.php [hunk 12]
```diff
@@ -31,7 +32,8 @@ public function __construct(
         private readonly int $maxValuesCount = 5000,
         private readonly \DateTime $minAllowedDate = new \DateTime('0000-01-01'),
         private readonly \DateTime $maxAllowedDate = new \DateTime('9999-12-31'),
-        private bool $supportForAttributes = true
+        private bool $supportForAttributes = true,
+        private readonly bool $supportUnsignedBigInt = true
```
Comment (abnegate) [inline src/Database/Validator/Query/Filter.php:36]: Any reason `supportForAttribute` is not a promoted property?

File: src/Database/Validator/Structure.php [hunk 13]
```diff
@@ -446,6 +461,11 @@ protected function checkForInvalidAttributeValues(array $structure, array $keys)
         return true;
     }
 
+    public function isBigIntStringWithinPhpIntRange(string $value, bool $signed): bool
```
Comment (abnegate) [inline src/Database/Validator/Structure.php]: I think we already have this method in the BigInt validator? If not, let's move it there anyway

File: src/Database/Database.php [hunk 14]
```diff
@@ -2521,6 +2524,7 @@ private function validateAttribute(
             filterCallback: fn ($id) => $this->adapter->filter($id),
             isMigrating: $this->isMigrating(),
             sharedTables: $this->getSharedTables(),
+            supportUnsignedBigInt: $this->adapter->getSupportForUnsignedBigInt(),
```
Comment (abnegate) [inline src/Database/Database.php]: Lets move `supportUnsignedBigInt` to after `supportForObject`

File: src/Database/Validator/Structure.php [hunk 15]
```diff
@@ -109,7 +109,8 @@ public function __construct(
         private readonly \DateTime $minAllowedDate = new \DateTime('0000-01-01'),
         private readonly \DateTime $maxAllowedDate = new \DateTime('9999-12-31'),
         private bool $supportForAttributes = true,
-        private readonly ?Document $currentDocument = null
+        private readonly ?Document $currentDocument = null,
+        private readonly bool $supportUnsignedBigInt = true
```
Comment (abnegate) [inline src/Database/Validator/Structure.php]: Let's flip the orderso supportX props are grouped

File: src/Database/Validator/Query/Filter.php (hunk 12 above)
Comment (ArnabChatterjee20k) [reply src/Database/Validator/Query/Filter.php:36]: @copilot can you address this

File: src/Database/Validator/Structure.php (hunk 13 above)
Comment (ArnabChatterjee20k) [reply src/Database/Validator/Structure.php]: @copilot can you address this

File: src/Database/Database.php (hunk 14 above)
Comment (ArnabChatterjee20k) [reply src/Database/Database.php]: @copilot can you address this

File: src/Database/Validator/Structure.php (hunk 15 above)
Comment (ArnabChatterjee20k) [reply src/Database/Validator/Structure.php]: @copilot can you address this

File: src/Database/Validator/Attribute.php [hunk 16]
```diff
@@ -522,6 +535,12 @@ protected function validateDefaultTypes(string $type, mixed $default): void
                     throw new DatabaseException($this->message);
                 }
                 break;
+            case Database::VAR_BIGINT:
+                if ($defaultType !== 'integer' && $defaultType !== 'string') {
+                    $this->message = 'Default value ' . $default . ' does not match given type ' . $type;
+                    throw new DatabaseException($this->message);
+                }
+                break;
```
[the remark this answers is not in the document]
Comment (ArnabChatterjee20k) [reply src/Database/Validator/Attribute.php:548]: fixed

File: src/Database/Adapter/Pool.php [hunk 17]
```diff
@@ -333,6 +333,11 @@ public function getLimitForInt(): int
         return $this->delegate(__FUNCTION__, \func_get_args());
     }
 
+    public function getLimitForBigInt(): int
+    {
+        return $this->delegate(__FUNCTION__, \func_get_args());
+    }
```
[the remark this answers is not in the document]
Comment (ArnabChatterjee20k) [reply src/Database/Adapter/Pool.php:359]: fixed

File: src/Database/Validator/Attribute.php [hunk 18]
```diff
@@ -420,6 +431,8 @@ public function checkType(Document $attribute): bool
                     Database::VAR_MEDIUMTEXT,
                     Database::VAR_LONGTEXT,
                     Database::VAR_INTEGER,
+                    Database::VAR_BIGINT,
```
Comment (abnegate) [inline src/Database/Validator/Attribute.php:434]:
```suggestion
```

**Review by abnegate — APPROVED** [review]
